### PR TITLE
feat: Port Datafusion Covariance to Comet

### PR DESCRIPTION
**Pull request by huaxingao**
## Which issue does this PR close?

Closes #.
Port `Covariance` in Comet. The reason we can't use DataFusion `Covariance` implementation is that Spark `Covariance` aggregate expression's [state](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Covariance.scala#L38) is different from DataFusion's [state](https://github.com/apache/arrow-datafusion/blob/main/datafusion/physical-expr/src/aggregate/covariance.rs#L91).

### CODE REVIEW

**Comment by huaxingao** [comment]
cc @andygrove @viirya

File: core/src/execution/datafusion/expressions/covariance.rs [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+    expressions::{format_state_name, StatsType},
+    AggregateExpr, PhysicalExpr,
+};
+
+/// COVAR and COVAR_SAMP aggregate expression
+#[derive(Debug, Clone)]
+pub struct Covariance {
+    name: String,
+    expr1: Arc<dyn PhysicalExpr>,
+    expr2: Arc<dyn PhysicalExpr>,
```
Comment (viirya) [inline core/src/execution/datafusion/expressions/covariance.rs:42]: Is there better names for these two children expressions?

File: core/src/execution/datafusion/expressions/covariance.rs [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+
+    fn field(&self) -> Result<Field> {
+        Ok(Field::new(&self.name, DataType::Float64, true))
+    }
+
+    fn create_accumulator(&self) -> Result<Box<dyn Accumulator>> {
+        Ok(Box::new(CovarianceAccumulator::try_new(
+            StatsType::Population,
+        )?))
+    }
```
Comment (viirya) [inline core/src/execution/datafusion/expressions/covariance.rs:160]: Looks like the difference between `CovariancePop` and `Covariance` is only the `StatsType` parameter when creating `CovarianceAccumulator`?

If so, maybe we only need `Covariance` and make `StatsType` as parameter when creating `Covariance`?

**Review by viirya — COMMENTED** [review]
We can update `EXPRESSIONS.md` too.

File: core/src/execution/datafusion/expressions/covariance.rs (hunk 2 above)
Comment (huaxingao) [reply core/src/execution/datafusion/expressions/covariance.rs:160]: https://github.com/apache/arrow-datafusion-comet/blob/955a0b9b9b026b477b5b527aba133d9e1402bd7e/core/src/execution/datafusion/expressions/covariance.rs#L368
Population Covariance is calculated over the entire dataset(N) whereas Sample Covariance is calculated over a sample (N-1)

Comment (viirya) [reply core/src/execution/datafusion/expressions/covariance.rs:160]: Yea, but the behavior is decided in `CovarianceAccumulator` based on its `StatsType`.

I mean this two struct `CovariancePop` and `Covariance`. They are basically the same, except that `name` and `StatsType` are different.

Comment (viirya) [reply core/src/execution/datafusion/expressions/covariance.rs:160]: Looks like we can simply have one `Covariance` and add one more parameter of `StatsType`.

File: core/src/execution/datafusion/expressions/covariance.rs (hunk 1 above)
Comment (huaxingao) [reply core/src/execution/datafusion/expressions/covariance.rs:42]: Spark uses `left` and `right`.

Comment (huaxingao) [reply core/src/execution/datafusion/expressions/covariance.rs:42]: I will leave this as is unless you prefer `left` and `right`.

File: core/src/execution/datafusion/expressions/covariance.rs (hunk 2 above)
Comment (huaxingao) [reply core/src/execution/datafusion/expressions/covariance.rs]: Changed. Thanks!

File: core/src/execution/datafusion/expressions/covariance.rs [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+use datafusion::logical_expr::Accumulator;
+use datafusion_common::{
+    downcast_value, unwrap_or_internal_err, DataFusionError, Result, ScalarValue,
+};
+use datafusion_physical_expr::{
+    aggregate::utils::down_cast_any_ref, expressions::format_state_name, AggregateExpr,
+    PhysicalExpr,
+};
+
+/// COVAR_SAMP and COVAR_POP aggregate expression
```
Comment (viirya) [inline core/src/execution/datafusion/expressions/covariance.rs:37]: It is better to mention why we need to port it in Comet.

File: core/src/execution/proto/expr.proto [hunk 4]
```diff
@@ -149,6 +151,20 @@ message BitXorAgg {
   DataType datatype = 2;
 }
 
+message CovSample {
+  Expr child1 = 1;
+  Expr child2 = 2;
+  bool null_on_divide_by_zero = 3;
```
Comment (viirya) [inline core/src/execution/proto/expr.proto:157]: We don't use this yet, right?

Comment (viirya) [reply core/src/execution/proto/expr.proto:157]: I also don't see it is specified in JVM side.

**Review by viirya — APPROVED** [review]
A few minor comments.

**Comment by viirya** [comment]
Merged. Thanks.

**Comment by huaxingao** [comment]
Thanks @viirya

**Pull request merged**